### PR TITLE
Refactor Stream/Task parsing

### PR DESCRIPTION
**Pull request by pperalta**
This is part of https://jira.spring.io/browse/XD-3377. Introduced `Tokens` data structure for holding tokens and the current token position used by parsers.

`TaskDslParser` and `StreamDslParser` were minimally modified; there is potential for further refactoring.

### CODE REVIEW

File: spring-cloud-data-core/src/main/java/org/springframework/cloud/data/core/dsl/Tokens.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+		Token t = tokenStream.get(position + distance);
+		return t.kind == desiredTokenKind;
+	}
+
+	/**
+	 * Return {@code true} if there are more tokens to process.
+	 *
+	 * @return {@code true} if there are more tokens to process
+	 */
+	protected boolean more() {
```
Comment (markfisher) [inline spring-cloud-data-core/src/main/java/org/springframework/cloud/data/core/dsl/Tokens.java]: maybe change to `hasNext()`? (or at least `hasMore()` but the latter is more consistent with an Iterator)

Comment (pperalta) [inline spring-cloud-data-core/src/main/java/org/springframework/cloud/data/core/dsl/Tokens.java]: Pushed

**Comment by markfisher** [comment]
This looks good to me. I'd like to get @aclement to review the general approach. Andy: the context is that we've migrated the syntactic parsing from XD so that it has no dependencies and can be used directly in the domain model. Also, we are drawing a clearer line between Stream and Task (i.e. Job), but we want to have reuse at the level of the parser (e.g. for an individual "module" and its key value pairs). The idea of introducing a `Tokens` datatype is to favor composition over inheritance for that lower level code.

**Comment by aclement** [comment]
Looks good to me. And glad to see there is no semantic checking in there, that was a mistake with the original form of the parser, it was too aware of information beyond the syntax. So if the two parsers do share a common definition of a 'module' the next step is to refactor `eatModule()` into something re-usable across both that operates on a `Tokens` object?

(Incidentally if you ever need it Spring Flo for XD uses a JavaScript version of the parser...)

**Comment by markfisher** [comment]
rebased, squashed, and merged as https://github.com/spring-cloud/spring-cloud-data/commit/4d9ae06086e39187039ff157a37010ff34fc72ad

**Pull request closed without merging**